lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1946 @@
+            wait_still_screen(10);
+        }
+        my $_temp = 1;
+        $_temp = script_run("timeout --kill-after=3 --signal=9 15 ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_addr ls");
+        $_ret |= $_temp;
+        record_info("Guest $_addr SSH PubKeyAuth failed", "Can not establish ssh connection to guest $_addr using Public Key Authentication", result => 'fail') if ($_temp != 0);
+    }
+    croak("SSH public key authentication setup failed for certain system") if ($_ret != 0 and $args{die} == 1);
+    return $_ret;
+}
+
+=head2 check_host_os
+
+Check source and destination hosts have the same architecture, otherwise test run
+can not proceed.
+=cut
+
+sub check_host_architecture {
+    my ($self, %args) = @_;
+
+    record_info("Check host architecture");
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_localarch = script_output("uname -i");
+    my $_peerip = get_var('PEER_IPADDR');
+    my $_peerarch = script_output("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip uname -i");
+    save_screenshot;
+    croak("Architecture $_localarch on $_localip does not match $_peerarch on $_peerip") if ($_localarch ne $_peerarch);
+}
+
+=head2 check_host_os
+
+Check source and destination hosts operating system version. Guest migration can
+not be done from the newer to the older. Main argument is host role (src or dst).
+=cut
+
+sub check_host_os {
+    my ($self, %args) = @_;
+    $args{role} //= 'src';
+
+    record_info("Check host os");
+    my $_ret = 0;
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    if (is_sle) {
+        my ($_localosver, $_localossp,) = get_os_release;
+        my ($_peerosver, $_peerossp,) = get_os_release("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip");
+        print "LOCAL OS $_localosver-sp$_localossp PEER OS $_peerosver-sp$_peerossp\n";
+        save_screenshot;
+        unless (($_peerosver > $_localosver) or ($_peerosver == $_localosver and $_peerossp >= $_localossp)) {
+            $_ret = 1;
+            if ($args{role} eq 'src') {
+                croak("Destination os $_peerosver-sp$_peerossp falls behind source os $_localosver-sp$_localossp");
+            }
+            elsif ($args{role} eq 'dst') {
+                record_info("Source os $_peerosver-sp$_peerossp  falls behind destination os $_localosver-sp$_localossp");
+            }
+        }
+    }
+    return $_ret;
+}
+
+=head2 check_host_virtualization
+
+Check virtualization modules and services are ready, otherwise test run can not
+proceed.
+=cut
+
+sub check_host_virtualization {
+    my $self = shift;
+
+    record_info("Check host virtualization");
+    if (is_kvm_host) {
+        assert_script_run("lsmod | grep kvm");
+    }
+    elsif (is_xen_host) {
+        assert_script_run("lsmod | grep xen");
+    }
+
+    if (script_run("systemctl is-active libvirtd") != 0) {
```

Comment:
This won't work for alp...

Reply:
You can skip for alp if needed later.

Reply:
Actually I was express similar things with https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/16509/files#r1132001780.

Reply:
I do not test on ALP, I can help add is_sle if you want. But I do not want to add something like is_alp or !is_alp, because I mainly do verification run on SLES.

Reply:
Got your point. It is ok on my side. Adding is_sle would work to tell where alp won't  fit and remind the designer the gap for multi-product automation support. This would save too much re-design work when extending it for TW/ALP. BTW, I am afraid Oliver would remind that this repo needs to have default for tumbleweed/leap. Better to pay attention too.